### PR TITLE
Implement OR match pattern

### LINKED ISSUE #769: Implement "or" match patterns in match expressions.
The "or" pattern is the one shown below:

```rust
let x = 1;
match x {
  0 | 1 => true,
  _ => false
}
```

### PR DESCRIPTION
**Pull request by IGI-111**
## Description

Allow users to specify match patters such as:

```sway
let x = 1;
match x {
  0 | 1 => true,
  _ => false
}
```

We also check that all patterns in a disjunction declare the same set of
variables (not doing so is an error).

Fix https://github.com/FuelLabs/sway/issues/769

This requires a change in the pattern matching analysis to remove the
assumption that a specialized matrix of a vector pattern is always a
vector, which is now no longer true because or patterns can generate
multiple branches and therefore multiple rows.

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

File: examples/match_statements/src/main.sw [hunk 1]
```diff
@@ -37,6 +37,11 @@ fn main(num: u64) -> u64 {
         Weather::Snowy => 20,
     };
 
+    let isSunny = match currentWeather {
```
Comment (JoshuaBatty) [inline examples/match_statements/src/main.sw:40]:
```suggestion
    let is_sunny = match current_weather {
```

File: sway-core/src/semantic_analysis/ast_node/expression/match_expression/analysis/witness_report.rs [hunk 2]
```diff
@@ -49,6 +49,7 @@ impl WitnessReport {
     ) -> CompileResult<(Pattern, Self)> {
         let mut warnings = vec![];
         let mut errors = vec![];
+        dbg!(&witness_report, &c);
```
Comment (JoshuaBatty) [inline sway-core/src/semantic_analysis/ast_node/expression/match_expression/analysis/witness_report.rs:52]: can this be removed?

File: examples/match_statements/src/main.sw (hunk 1 above)
Comment (IGI-111) [reply examples/match_statements/src/main.sw:40]: I'd rather follow the convention within the file, but maybe we should change the whole file to use snake case for identifiers.

Comment (JoshuaBatty) [reply examples/match_statements/src/main.sw:40]: my vote is to change all occurrences to snake case.

**Review by JoshuaBatty — APPROVED** [review]

**Pull request merged**